src/clients/arclib/arcsub.1.in
.\" -*- nroff -*-
.TH ARCSUB 1 "@DATE@" "NorduGrid ARC @VERSION@" "NorduGrid Users Manual"

.SH NAME

arcsub \- ARC Submission

.SH DESCRIPTION

The
.B arcsub
command is used for submitting jobs to grid enabled computing resources.

.SH SYNOPSIS

.B arcsub [options] [filename ...]

.SH OPTIONS

.IP "\fB-c\fR, \fB--cluster\fR=\fIname\fR"
select one or more computing elements: \fIname\fR can be an alias for a single CE, a group of CEs or a URL
.IP "\fB-g\fR, \fB--index\fR=\fIname\fR"
select one or more registries: \fIname\fR can be an alias for a single registry, a group of registries or a URL
.IP "\fB-R\fR, \fB--rejectdiscovery\fR=\fIURL\fR"
skip the service with the given URL during service discovery
.IP "\fB-S\fR, \fB--submissioninterface\fR=\fIInterfaceName\fR"
only use this interface for submitting (e.g. org.nordugrid.gridftpjob, org.ogf.glue.emies.activitycreation, org.ogf.bes)
.IP "\fB-I\fR, \fB--infointerface\fR=\fIInterfaceName\fR"
the computing element specified by URL at the command line should be queried using this information interface
(possible options: org.nordugrid.ldapng, org.nordugrid.ldapglue2, org.nordugrid.wsrfglue2, org.ogf.glue.emies.resourceinfo)
.IP "\fB-e\fR, \fB--jobdescrstring=\fIString"
jobdescription string describing the job to be submitted
.IP "\fB-f\fR, \fB--jobdescrfile=\fIfilename"
jobdescription file describing the job to be submitted
.IP "\fB-j\fR, \fB--joblist\fR=\fIfilename\fR"
the file storing information about active jobs (default ~/.arc/jobs.xml)
.IP "\fB-o\fR, \fB--jobids-to-file\fR=\fIfilename\fR"
the IDs of the submitted jobs will be appended to this file
.IP "\fB-D\fR, \fB--dryrun\fR"
submit jobs as dry run (no submission to batch system)
.IP "\fB    --direct\fR"
submit directly - no resource discovery or matchmaking
.IP "\fB-x\fR, \fB--dumpdescription\fR"
do not submit - dump job description in the language accepted by the target
.IP "\fB-P\fR, \fB--listplugins\fR"
list the available plugins
.IP "\fB-t\fR, \fB--timeout\fR=\fIseconds\fR"
timeout in seconds (default 20)
.IP "\fB-z\fR, \fB--conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB-d\fR, \fB--debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB-b\fR, \fB--broker\fR=\fIbroker\fR"
selected broker: Random (default), FastestQueue or custom
.IP "\fB-v\fR, \fB--version\fR"
print version information
.IP "\fB-?\fR, \fB--help\fR"
print help
.LP

.SH ARGUMENTS

.IP "\fBfilename\fR ..."
job description files describing the jobs to be submitted
.LP

.SH EXTENDED DESCRIPTION

\fBarcsub\fR is the key command when submitting jobs to grid enabled computing
resources with the ARC client. As default \fBarcsub\fR is able to submit jobs to
A-REX and CREAM computing elements (CEs), and as always for successful
submission you need to be authenticated at the targeted computing services. Since
\fBarcsub\fR is build on a modular library, modules can be installed which
enables submission to other targets, e.g. the classic ARC CE Grid-Manager.

Job submission can be accomplished by invoking the \fBarcsub\fR command
with the \fB--cluster\fR option set to the target CE (URL, alias or group), and the
job description file given as argument.
Multiple CEs can be targeted for submission by specifying the
\fB--cluster\fR option multiple times, in which case the CEs will be ranked by
the resource broker. If no \fB--cluster\fR or \fB--index\fR (see below) options
have been specified, then the defaults, which is read from the configuration
file (see the files section), will be used. That also makes it possible to
define own defaults, so it makes it superfluous to specify the CE usually used
for submission. In the same manner aliases and groups, defined in the configuration file,
can be utilized, and can be used as argument to the \fB--cluster\fR or
\fB--index\fR options.

The format of a classic GRIDFTP-based cluster URLs:
.br
\fB[ldap://]<hostname>[:2135/nordugrid-cluster-name=<hostname>,Mds-Vo-name=local,o=grid]\fR
.br
Only the \fBhostname\fR part has to be specified, the rest of the URL is automatically generated.

The format of an A-REX URL is:
.br
\fB[https://]<hostname>[:<port>][/<path>]\fR
.br
Here the port is 443 by default, but the path cannot be guessed, so if it is not specified, then
the service is assumed to live on the root path.

Job descriptions can also be specified using the \fB--jobdescrfile\fR option
which expect the file name of the description as argument, or the
\fB--jobdescrstring\fR option which expect as argument the job description as
a string, and both options can be specified multiple times and one does not
exclude the other. The default supported job description languages are XRSL,
JSDL and JDL.

If the job description is successfully submitted a job-ID is returned and
printed. This job-ID uniquely identifies the job while it is being executed.
On the other hand it is also possible that no CEs matches the constraints
defined in the description in which case no submission will be done. Upon
successful submission, the job-ID along with more technical job information is
stored in the job-list file (described below). The stored information enables
the job management commands of the ARC client to manage jobs easily, and thus
the job-ID need not to be saved manually. By default the job-list file is
stored in the .arc directory in the home directory of the user, however another
location can be specified using the \fB--joblist\fR option taking the location
of this file as argument. If the \fB--joblist\fR option was used during
submission, it should also be specified in the consecutive commands when
managing the job. If a Computing Element has multiple job submission interfaces
(e.g. gridftp, EMI-ES, BES), then the brokering algorithm will choose one of them.
With the \fB--submissioninterface\fR option the requested interface
can be specified, and in that case only those Computing Elements will be
considered which has that specific interface, and only that interface
will be used to submit the jobs.

Instead of directly targeting a CE for submission, using the \fB--cluster\fR
option, an information index server can be specified with the \fB--index\fR
option. Specifying one or multiple index servers instructs the \fBarcsub\fR
command to query the servers for registered CEs, the returned CEs will then be
matched against the job description and those matching will be ranked by the
chosen broker (see below) and submission will be tried in order until
successful or reaching the end.
From the returned list of CEs it might happen that a troublesome or
undesirable CE is selected for submission, in that case it possible to reject
that cluster using the \fB--rejectdiscovery\fR option and providing the URL
(or just the hostname) of the CE, which will disregard that CE as a target for submission.

When multiple CEs are targeted for submission, the resource broker will be used
to rank the CEs. The broker used by default will rank the CEs randomly,
however a different broker can be chosen by using the \fB--broker\fR option,
which takes the name of the broker as argument. Two other brokers are
by default available. The FastestQueue and Benchmark brokers. The former ranks
CEs according to the length of the job queue at the CEs, ranking those with
shortest queue first/highest. The latter ranks CEs according to a specified
benchmark, which should be specified by appending the broker name with ':' and
then the name of the benchmark. If no option is given to the Benchmark broker
then CEs will be ranked according to the 'specint2000' benchmark. For examples
see the respective section.

Before submission, \fBarcsub\fR performs an intelligent modification of the job
description (adding or modifying attributes, even converting the description
language to fit the needs of the CE) ensuring that it is valid. The
modified job description can be printed by specifying the
\fB--dumpdescription\fR option. The format, i.e. job description language, of
the printed job description cannot be specified, and will be that which will
be sent to and accepted by the chosen target. Further information from
\fBarcsub\fR can be obtained by increasing the verbosity, which is done with
the \fB--debug\fR option where the default verbosity level is WARNING. Setting
the level to DEBUG will show all messages, while setting it to FATAL will
only show fatal log messages.

To \fBvalidate\fR your job description without actually submitting a job, use
the \fB--dryrun\fR option: it will capture possible syntax or other errors, but
will instruct the site not to submit the job for execution. Only the
grid-manager (ARC0) and A-REX (ARC1) CEs support this feature.

.SH EXAMPLES

Submission of a job description file "helloworld.jsdl" to ce.example.com:
.br
\fBarcsub -c ce.example.com helloworld.jsdl\fR

A information index server (registry) can also be queried for CEs to submit to:
.br
\fBarcsub -g registry.example.com helloworld.jsdl\fR

The job description can also be specified directly on the command line as shown
in the example, using the XRSL job description language:
.br
\fBarcsub -c example.com/arex -e \\
.br
\'&(executable="/bin/echo")(arguments="Hello World!")\'\fR

When submitting against CEs retrieved from information index servers it might
be useful to do resource brokering:
.br
\fBarcsub -g registry.example.com -b FastestQueue helloworld.jsdl\fR

Disregarding a specific CE for submission submitting against an information
index server:
.br
\fBarcsub -g registry.example.com -R badcomputingelement.com/arex helloworld.jsdl\fR

Dumping the job description is done as follows:
.br
\fBarcsub -c example.com/arex -x helloworld.jsdl\fR

.SH FILES

.TP
.B ~/.arc/client.conf
Some options can be given default values by specifying them in the ARC client
configuration file. Registry and computing element services can be specified
in seperate sections of the config. The default services can be specified by
adding 'default=yes' attribute to the section of the service,
thus when no \fB--cluster\fR or \fB--index\fR
options are given these will be used for submission. Each service has an alias,
and can be member of any number of groups. Then specifying the alias or the
name of the group with the \fB--cluster\fR or \fB--index\fR options will
select the given services.
By using the \fB--conffile\fR option a different configuration file can be used
than the default. Note that some installations also have a system client
configuration file, however attributes in the client one takes precedence,
and then command line options takes precedence over configuration file
attributes.

.TP
.B ~/.arc/jobs.xml
This a local list of the user's active jobs. When a job is successfully
submitted it is added to this list and when it is removed from the remote
cluster it is removed from this list. This list is used as the list of all
active jobs when the user specifies the \fB--all\fR option to the various
NorduGrid ARC user interface commands. By using the \fB--joblist\fR option a
different file can be used than the default.

.SH ENVIRONMENT VARIABLES

.TP
.B X509_USER_PROXY
The location of the user's grid proxy file. Shouldn't be set unless
the proxy is in a non-standard location.

.TP
.B ARC_LOCATION
The location where ARC is installed can be specified by this variable. If not
specified the install location will be determined from the path to the
command being executed, and if this fails a WARNING will be given stating the
location which will be used.

.TP
.B ARC_PLUGIN_PATH
The location of ARC plugins can be specified by this variable. Multiple
locations can be specified by separating them by : (; in Windows). The
default location is \fB$ARC_LOCATION\fR/lib/arc (\\ in Windows).

.SH COPYRIGHT

APACHE LICENSE Version 2.0

.SH AUTHOR

Mattias Ellert <mattias.ellert@fysast.uu.se>

.SH SEE ALSO

.BR arccat (1),
.BR arcclean (1),
.BR arccp (1),
.BR arcget (1),
.BR arcinfo (1),
.BR arckill (1),
.BR arcls (1),
.BR arcmigrate (1),
.BR arcmkdir (1),
.BR arcproxy (1),
.BR arcrenew (1),
.BR arcresub (1),
.BR arcresume (1),
.BR arcrm (1),
.BR arcstat (1),
.BR arcsync (1),
.BR arctest (1)



NEWS
no news good news :)


src/clients/arclib/arcsub.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <list>
#include <string>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#include <arc/ArcLocation.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/UserConfig.h>
#include <arc/Utils.h>
#include <arc/compute/Broker.h>
#include <arc/compute/ComputingServiceRetriever.h>
#include <arc/compute/ExecutionTarget.h>
#include <arc/compute/Job.h>
#include <arc/compute/JobDescription.h>
#include <arc/compute/SubmissionStatus.h>
#include <arc/compute/Submitter.h>

#include "utils.h"

static Arc::Logger logger(Arc::Logger::getRootLogger(), "arcsub");

static int submit(const Arc::UserConfig& usercfg, const std::list<Arc::JobDescription>& jobdescriptionlist, const std::list<Arc::Endpoint>& services, const std::list<std::string> requestedSubmissionInterfaces, const std::string& jobidfile, bool direct_submission);
static int dumpjobdescription(const Arc::UserConfig& usercfg, const std::list<Arc::JobDescription>& jobdescriptionlist, const std::list<Arc::Endpoint>& services, const std::list<std::string> requestedSubmissionInterfaces);

int RUNMAIN(arcsub)(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::getRootLogger().setThreshold(Arc::WARNING);

  Arc::ArcLocation::Init(argv[0]);

  ClientOptions opt(ClientOptions::CO_SUB,
                    istring("[filename ...]"),
                    istring("The arcsub command is used for "
                            "submitting jobs to Grid enabled "
                            "computing\nresources."));

  std::list<std::string> params = opt.Parse(argc, argv);

  if (opt.showversion) {
    std::cout << Arc::IString("%s version %s", "arcsub", VERSION)
              << std::endl;
    return 0;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!opt.debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(opt.debug));

  Arc::UserConfig usercfg(opt.conffile, opt.joblist);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (opt.show_plugins) {
    std::list<std::string> types;
    types.push_back("HED:SubmitterPlugin");
    types.push_back("HED:ServiceEndpointRetrieverPlugin");
    types.push_back("HED:TargetInformationRetrieverPlugin");
    types.push_back("HED:JobDescriptionParser");
    types.push_back("HED:BrokerPlugin");
    showplugins("arcsub", types, logger, usercfg.Broker().first);
    return 0;
  }

  if (!checkproxy(usercfg)) {
    return 1;
  }

  if (opt.debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(usercfg.Verbosity()));

  if (opt.timeout > 0)
    usercfg.Timeout(opt.timeout);

  if (!opt.broker.empty())
    usercfg.Broker(opt.broker);

  opt.jobdescriptionfiles.insert(opt.jobdescriptionfiles.end(),
                                 params.begin(), params.end());

  if (opt.jobdescriptionfiles.empty() && opt.jobdescriptionstrings.empty()) {
    logger.msg(Arc::ERROR, "No job description input specified");
    return 1;
  }

  std::list<Arc::JobDescription> jobdescriptionlist;

  // Loop over input job description files
  for (std::list<std::string>::iterator it = opt.jobdescriptionfiles.begin();
       it != opt.jobdescriptionfiles.end(); it++) {

    std::ifstream descriptionfile(it->c_str());

    if (!descriptionfile) {
      logger.msg(Arc::ERROR, "Can not open job description file: %s", *it);
      return 1;
    }

    descriptionfile.seekg(0, std::ios::end);
    std::streamsize length = descriptionfile.tellg();
    descriptionfile.seekg(0, std::ios::beg);

    char *buffer = new char[length + 1];
    descriptionfile.read(buffer, length);
    descriptionfile.close();

    buffer[length] = '\0';
    std::list<Arc::JobDescription> jobdescs;
    if (Arc::JobDescription::Parse((std::string)buffer, jobdescs)) {
      for (std::list<Arc::JobDescription>::iterator itJ = jobdescs.begin();
           itJ != jobdescs.end(); itJ++) {
        itJ->Application.DryRun = opt.dryrun;
        for (std::list<Arc::JobDescription>::iterator itJAlt = itJ->GetAlternatives().begin();
             itJAlt != itJ->GetAlternatives().end(); itJAlt++) {
          itJAlt->Application.DryRun = opt.dryrun;
        }
      }

      jobdescriptionlist.insert(jobdescriptionlist.end(), jobdescs.begin(), jobdescs.end());
    }
    else {
      logger.msg(Arc::ERROR, "Invalid JobDescription:");
      std::cout << buffer << std::endl;
      delete[] buffer;
      return 1;
    }
    delete[] buffer;
  }

  //Loop over job description input strings
  for (std::list<std::string>::iterator it = opt.jobdescriptionstrings.begin();
       it != opt.jobdescriptionstrings.end(); it++) {

    std::list<Arc::JobDescription> jobdescs;
    if (Arc::JobDescription::Parse(*it, jobdescs)) {
      for (std::list<Arc::JobDescription>::iterator itJ = jobdescs.begin();
           itJ != jobdescs.end(); itJ++) {
        itJ->Application.DryRun = opt.dryrun;
        for (std::list<Arc::JobDescription>::iterator itJAlt = itJ->GetAlternatives().begin();
             itJAlt != itJ->GetAlternatives().end(); itJAlt++) {
          itJAlt->Application.DryRun = opt.dryrun;
        }
      }

      jobdescriptionlist.insert(jobdescriptionlist.end(), jobdescs.begin(), jobdescs.end());
    }
    else {
      logger.msg(Arc::ERROR, "Invalid JobDescription:");
      std::cout << *it << std::endl;
      return 1;
    }
  }

  std::list<Arc::Endpoint> services = getServicesFromUserConfigAndCommandLine(usercfg, opt.indexurls, opt.clusters, opt.requestedSubmissionInterfaceName, opt.infointerface);

  if (!opt.direct_submission) {
    usercfg.AddRejectDiscoveryURLs(opt.rejectdiscovery);
  }

  std::list<std::string> rsi;
  if(!opt.requestedSubmissionInterfaceName.empty()) rsi.push_back(opt.requestedSubmissionInterfaceName);

  if (opt.dumpdescription) {
    return dumpjobdescription(usercfg, jobdescriptionlist, services, rsi);
  }

  return submit(usercfg, jobdescriptionlist, services, rsi, opt.jobidoutfile, opt.direct_submission);
}


class HandleSubmittedJobs : public Arc::EntityConsumer<Arc::Job> {
public:
  HandleSubmittedJobs(const std::string& jobidfile, const std::string& joblist) : jobidfile(jobidfile), joblist(joblist), submittedJobs() {}

  void addEntity(const Arc::Job& j) {
    std::cout << Arc::IString("Job submitted with jobid: %s", j.JobID) << std::endl;
    submittedJobs.push_back(j);
  }
  
  void write() const {
    if (!jobidfile.empty() && !Arc::Job::WriteJobIDsToFile(submittedJobs, jobidfile)) {
      logger.msg(Arc::WARNING, "Cannot write jobids to file (%s)", jobidfile);
    }
    if (!Arc::Job::WriteJobsToFile(joblist, submittedJobs)) {
      std::cout << Arc::IString("Warning: Failed to lock job list file %s", joblist)
                << std::endl;
      std::cout << Arc::IString("To recover missing jobs, run arcsync") << std::endl;
    }
  }

  void printsummary(const std::list<Arc::JobDescription>& originalDescriptions, const std::list<const Arc::JobDescription*>& notsubmitted) const {
    if (originalDescriptions.size() > 1) {
      std::cout << std::endl << Arc::IString("Job submission summary:") << std::endl;
      std::cout << "-----------------------" << std::endl;
      std::cout << Arc::IString("%d of %d jobs were submitted", submittedJobs.size(), submittedJobs.size()+notsubmitted.size()) << std::endl;
      if (!notsubmitted.empty()) {
        std::cout << Arc::IString("The following %d were not submitted", notsubmitted.size()) << std::endl;
        for (std::list<const Arc::JobDescription*>::const_iterator it = notsubmitted.begin();
             it != notsubmitted.end(); ++it) {
          int jobnr = 1;
          for (std::list<Arc::JobDescription>::const_iterator itOrig = originalDescriptions.begin();
               itOrig != originalDescriptions.end(); ++itOrig, ++jobnr) {
            if (&(*itOrig) == *it) {
              std::cout << Arc::IString("Job nr.") << " " << jobnr;
              if (!(*it)->Identification.JobName.empty()) {
                std::cout << ": " << (*it)->Identification.JobName;
              }
              std::cout << std::endl;
              break;
            }
          }
        }
      }
    }
  }

  void clearsubmittedjobs() { submittedJobs.clear(); }

private:
  const std::string jobidfile, joblist;
  std::list<Arc::Job> submittedJobs;
};


static bool match_submission_interface(const Arc::ExecutionTarget& target, const std::list<std::string>& requestedSubmissionInterfaces) {
  if(requestedSubmissionInterfaces.empty()) return true;
  for(std::list<std::string>::const_iterator iname = requestedSubmissionInterfaces.begin();
                iname != requestedSubmissionInterfaces.end();++iname) {
    if(*iname == target.ComputingEndpoint->InterfaceName) return true;
  }
  return false;
}

static int submit(const Arc::UserConfig& usercfg, const std::list<Arc::JobDescription>& jobdescriptionlist, const std::list<Arc::Endpoint>& services, const std::list<std::string> requestedSubmissionInterfaces, const std::string& jobidfile, bool direct_submission) {
  int retval = 0;
  
  HandleSubmittedJobs hsj(jobidfile, usercfg.JobListFile());
  Arc::Submitter s(usercfg);
  s.addConsumer(hsj);

  Arc::SubmissionStatus status;
  if (!direct_submission) {
    status = s.BrokeredSubmit(services, jobdescriptionlist, requestedSubmissionInterfaces);
  }
  else {
    status = s.Submit(services, jobdescriptionlist);
  }
  hsj.write();

  if (status.isSet(Arc::SubmissionStatus::BROKER_PLUGIN_NOT_LOADED)) {
    std::cerr << Arc::IString("ERROR: Unable to load broker %s", usercfg.Broker().first) << std::endl;
    return 1;
  }
  if (status.isSet(Arc::SubmissionStatus::NO_SERVICES)) {
   std::cerr << Arc::IString("ERROR: Job submission aborted because no resource returned any information") << std::endl;
   return 1;
  }
  if (status.isSet(Arc::SubmissionStatus::SUBMITTER_PLUGIN_NOT_LOADED)) {
    bool gridFTPJobPluginFailed = false;
    for (std::map<Arc::Endpoint, Arc::EndpointSubmissionStatus>::const_iterator it = s.GetEndpointSubmissionStatuses().begin();
         it != s.GetEndpointSubmissionStatuses().end(); ++it) {
      if (it->first.InterfaceName == "org.nordugrid.gridftpjob" && it->second == Arc::EndpointSubmissionStatus::NOPLUGIN) {
        gridFTPJobPluginFailed = true;
      }
    }
    if (gridFTPJobPluginFailed) {
      std::cerr << Arc::IString("ERROR: A computing resource using the gridftp interface was requested, but") << std::endl;
      std::cerr << Arc::IString("       the corresponding plugin could not be loaded. Is the plugin installed?") << std::endl;
      std::cerr << Arc::IString("       If not, please install the package 'nordugrid-arc-plugins-globus'.") << std::endl;
      std::cerr << Arc::IString("       Depending on your type of installation the package name might differ. ") << std::endl;
    }
    // TODO: What to do when failing to load other plugins.
    retval = 1;
  }
  if (status.isSet(Arc::SubmissionStatus::DESCRIPTION_NOT_SUBMITTED)) {
    std::cerr << Arc::IString("ERROR: One or multiple job descriptions was not submitted.") << std::endl;
    retval = 1;
  }
    
  hsj.printsummary(jobdescriptionlist, s.GetDescriptionsNotSubmitted());

  return retval;
}

static int dumpjobdescription(const Arc::UserConfig& usercfg, const std::list<Arc::JobDescription>& jobdescriptionlist, const std::list<Arc::Endpoint>& services, std::list<std::string> requestedSubmissionInterfaces) {
  int retval = 0;

  std::set<std::string> preferredInterfaceNames;
  if (usercfg.InfoInterface().empty()) {
    preferredInterfaceNames.insert("org.nordugrid.ldapglue2");
  } else {
    preferredInterfaceNames.insert(usercfg.InfoInterface());
  }

  Arc::ComputingServiceUniq csu;
  Arc::ComputingServiceRetriever csr(usercfg, std::list<Arc::Endpoint>(), usercfg.RejectDiscoveryURLs(), preferredInterfaceNames);
  csr.addConsumer(csu);
  for (std::list<Arc::Endpoint>::const_iterator it = services.begin(); it != services.end(); it++) {
    csr.addEndpoint(*it);
  }
  csr.wait();
  std::list<Arc::ComputingServiceType> CEs = csu.getServices();


  if (CEs.empty()) {
    std::cout << Arc::IString("Dumping job description aborted because no resource returned any information") << std::endl;
    return 1;
  }

  Arc::Broker broker(usercfg, usercfg.Broker().first);
  if (!broker.isValid(false)) {
    logger.msg(Arc::ERROR, "Dumping job description aborted: Unable to load broker %s", usercfg.Broker().first);
    return 1;
  }

  Arc::ExecutionTargetSorter ets(broker, CEs);
  std::list<Arc::JobDescription>::const_iterator itJAlt; // Iterator to use for alternative job descriptions.
  for (std::list<Arc::JobDescription>::const_iterator itJ = jobdescriptionlist.begin();
       itJ != jobdescriptionlist.end(); ++itJ) {
    const Arc::JobDescription* currentJobDesc = &*itJ;
    bool descriptionDumped = false;
    do {
      Arc::JobDescription jobdescdump(*currentJobDesc);
      ets.set(jobdescdump);

      for (ets.reset(); !ets.endOfList(); ets.next()) {
        if(!match_submission_interface(*ets,requestedSubmissionInterfaces)) continue;
        if (!jobdescdump.Prepare(*ets)) {
          logger.msg(Arc::INFO, "Unable to prepare job description according to needs of the target resource (%s).", ets->ComputingEndpoint->URLString); 
          continue;
        }
  
        std::string jobdesclang = "nordugrid:jsdl";
        if (ets->ComputingEndpoint->InterfaceName == "org.nordugrid.gridftpjob") {
          jobdesclang = "nordugrid:xrsl";
        }
        else if (ets->ComputingEndpoint->InterfaceName == "org.glite.cream") {
          jobdesclang = "egee:jdl";
        }
        else if (ets->ComputingEndpoint->InterfaceName == "org.ogf.glue.emies.activitycreation") {
          jobdesclang = "emies:adl";
        }
        std::string jobdesc;
        if (!jobdescdump.UnParse(jobdesc, jobdesclang)) {
          logger.msg(Arc::INFO, "An error occurred during the generation of job description to be sent to %s", ets->ComputingEndpoint->URLString); 
          continue;
        }
  
        std::cout << Arc::IString("Job description to be sent to %s:", ets->AdminDomain->Name) << std::endl;
        std::cout << jobdesc << std::endl;
        descriptionDumped = true;
        break;
      }

      if (!descriptionDumped && itJ->HasAlternatives()) { // Alternative job descriptions.
        if (currentJobDesc == &*itJ) {
          itJAlt = itJ->GetAlternatives().begin();
        }
        else {
          ++itJAlt;
        }
        currentJobDesc = &*itJAlt;
      }
    } while (!descriptionDumped && itJ->HasAlternatives() && itJAlt != itJ->GetAlternatives().end());

    if (ets.endOfList()) {
      std::cout << Arc::IString("Unable to prepare job description according to needs of the target resource.") << std::endl;
      retval = 1;
    }
  } //end loop over all job descriptions

  return retval;
}
